ocaml/xapi/rrdd_proxy.ml

```diff
@@ -97,15 +99,25 @@ let get_vm_rrd_forwarder (req : Http.Request.t) (s : Unix.file_descr) _ =
           let owner = Db.VM.get_resident_on ~__context ~self:vm_ref in
           let owner_uuid = Db.Host.get_uuid ~__context ~self:owner in
           let is_owner_localhost = owner_uuid = localhost_uuid in
-          if is_owner_localhost then
-            if is_master then
+          let owner_is_available =
+            is_owner_online owner && not is_xapi_initialising
+          in
+          match
+            (Pool_role.get_role (), is_owner_localhost, owner_is_available)
+          with
+          | (Master | Slave _), false, true ->
+              (* VM is running elsewhere *)
+              read_at_owner owner
+          | Master, true, _ | Master, false, false ->
+              (* VM running on node, or not running at all. *)
               unarchive ()
-            else
+          | Slave _, true, _ | Slave _, _, false ->
+              (* Coordinator knows best *)
               unarchive_at_master ()
-          else if is_owner_online owner && not is_xapi_initialising then
-            read_at_owner owner
-          else
-            unarchive_at_master ()
+          | Broken, _, _ ->
+              info "%s: host is broken, VM's metrics are not available"
+                __FUNCTION__ ;
+              unavailable ()
```

Comment:
Is 503 the right code, or should it be 502 or 504? We are proxying the call somewhere else, but the "somewhere else" has failed.

Reply:
The client is sending the request to this code. The fact that it's actually a proxy is incidental, so I think 503 is an appropriate response.

Looking at the other two:
> The HyperText Transfer Protocol (HTTP) 502 Bad Gateway server error response code indicates that the server, while acting as a gateway or proxy, received an invalid response from the upstream server.

I don't think the description applies to the situation, as this code hasn't made calls to upstream servers yet.

> The HyperText Transfer Protocol (HTTP) 504 Gateway Timeout server error response code indicates that the server, while acting as a gateway or proxy, did not get a response in time from the upstream server that it needed in order to complete the request.

Again, since there's no request sent to an upstream server, this HTTP code is not relevant.